### PR TITLE
Add calls to process headers as well

### PR DESCRIPTION
**Pull request by ponychicken**
For an archiving project I needed to be able to access all the headers of the mails as well.
This commit adds the string containing the original headers to the messages

### CODE REVIEW

File: lib/yugo/client.ex [hunk 1]
```diff
@@ -399,6 +399,7 @@ defmodule Yugo.Client do
           |> Enum.reject(fn {key, _} -> Map.has_key?(msg, key) end)
           |> Enum.map(&elem(&1, 1))
 
+        parts_to_fetch = ["RFC822.HEADER" | parts_to_fetch]
         parts_to_fetch = ["BODY" | parts_to_fetch]
```
Comment (Flying-Toast) [inline lib/yugo/client.ex:403]: This can be `parts_to_fetch = ["RFC822.HEADER", "BODY" | parts_to_fetch]`

**Comment by Flying-Toast** [comment]
Can you please update the tests too? We should also add some new tests specifically for headers.

**Comment by ponychicken** [comment]
I have fixed the line and added two tests to the parser

**Comment by Flying-Toast** [comment]
Did you push? It doesn't look like the commits in the PR updated.

**Comment by tzugen** [comment]
Ah yes the push failed and I closed my laptop. Its there now.

**Comment by Flying-Toast** [comment]
Great. Just a few more things:
    1) Let’s update the `Yugo.email` typespec to include the new field - I also think we can name the field simply “headers” instead of “rfc822_header”
    2) I see you updated the expected client messages in the tests to expect the client to *request* the RFC822.HEADER field, but it still doesn’t have the server send them back. You should change the `S: …` lines so that they include a variety of sample headers and then ensure that the email that is `receive`d actually contains those headers.
    3) I’d like us to parse the headers into a more elixir-y data type (probably an array of `{name, value}` tuples). This might be as simple as splitting each line on “: “ but check the IMAP spec to see what exactly it would entail.

Thanks again for the work so far!

**Comment by ponychicken** [comment]
Ok, 1 & 3 are done.
Regarding 2 I am not sure how to proceed.

A sample response might be this:
`S: * 2 FETCH (FLAGS (\sEEn) BODY ("text" "plain" ("charset" "us-ascii" "format" "flowed") NIL NIL "7bit" 47 6) RFC822.HEADER {1992}\r\nReturn-Path: <email@example.com>\r\n ENVELOPE ("Wed, 07 Dec 2022 18:02:41 -0500" NIL (("Marge Simpson" NIL "marge" "simpsons-family.com")) (("Marge Simpson" NIL "marge" "simpsons-family.com")) (("Marge" NIL "marge" "simpsons-family.com")) (("HOMIEEEE" NIL "homer" "simpsons-family.com")) NIL NIL NIL "fjaelwkjfi oaf<$ ))) \""))`

Right now the assert_comms are wrapped in `~S` sigil, which means it will double escape the line breaks.
But if I use the `~s`, to keep the linebreaks in the header sample, assert_comms will break, because it splits on `\n` and suddenly there will be `\r\n` in the middle of the string.

**Comment by Flying-Toast** [comment]
I'm not sure I understand the issue. `assert_comms()` replaces all "\n" with "\r\n" then splits on the regex `~r/\r\n[CS]:/`. You should be able to put literal line breaks in the string, as is done [here](https://github.com/Flying-Toast/yugo/blob/3fd2b84e8dbaac739cddb851ed5f6be5541f3c80/test/yugo/client_test.exs#L69). Unless I am missing something - did you try that and it didn't work?

File: lib/yugo.ex [hunk 2]
```diff
@@ -46,6 +46,7 @@ defmodule Yugo do
           cc: [address],
           date: DateTime.t(),
           flags: [flag],
+          headers: list(list(String.t())),
```
Comment (Flying-Toast) [inline lib/yugo.ex:49]: Can you make this `headers: [{String.t(), String.t()}]`

File: lib/yugo/parser.ex [hunk 3]
```diff
@@ -225,6 +225,15 @@ defmodule Yugo.Parser do
       name == "BODY" ->
         parse_body(rest)
 
+      name == "RFC822.HEADER" ->
+        {raw_headers, rest} = parse_string(rest)
+
+        {{:headers,
+          raw_headers
+            |> String.split("\r\n", trim: true)
+            |> Enum.map(&String.split(&1, ":", trim: true))
```
Comment (Flying-Toast) [inline lib/yugo/parser.ex:234]: this breaks if the value has a ":" in it, e.g. `Some-Header: a:b:c:d`

File: test/yugo/parser_test.exs [hunk 4]
```diff
@@ -82,4 +82,34 @@ defmodule Yugo.ParserTest do
     [fetch: {0, :flags, ["\\Seen", "\\Recent"]}, fetch: {0, :uid, 84}] =
       Parser.parse_response(~S|* 0 fetch (UID 84 FLags (\Seen \Recent))|)
   end
+
+  test "parse FETCH response with simple RFC822.HEADER" do
+    [
+      fetch: {1, :headers, [["From", " example@example.com"], ["Subject", " Test Email"]]}
```
Comment (Flying-Toast) [inline test/yugo/parser_test.exs:88]: The values shouldn't all have leading spaces

File: test/yugo/parser_test.exs [hunk 5]
```diff
@@ -82,4 +82,34 @@ defmodule Yugo.ParserTest do
...
+              "X-Spam-Checker-Version: SpamAssassin 3.4.6 (2021-04-09)\r\n" <>
+              "X-Spam-Level:\r\n" <>
+              "X-Spam-Pyzor:\r\n" <>
+              "X-Spam-Status: No, score=-100.0 required=6.0 shortcircuit=ham
+	autolearn=disabled version=3.4.6\r\n" <>
+              "Subject: Multi-Line Subject\r\n\t Continued Here\r\n" <>
+              "To: recipient1@example.com,\r\n recipient2@example.com\r\n" <>
+              "CC: cc@example.com\r\nBCC: bcc@example.com\r\n\r\n"
+    [
+      fetch: {2, :headers, [["From", " example@example.com"], ["X-Spam-Checker-Version", " SpamAssassin 3.4.6 (2021-04-09)"], ["X-Spam-Level"], ["X-Spam-Pyzor"], ["X-Spam-Status", " No, score=-100.0 required=6.0 shortcircuit=ham\n\tautolearn=disabled version=3.4.6"], ["Subject", " Multi-Line Subject"], ["\t Continued Here"], ["To", " recipient1@example.com,"], [" recipient2@example.com"], ["CC", " cc@example.com"], ["BCC", " bcc@example.com"]]}
```
Comment (Flying-Toast) [inline test/yugo/parser_test.exs:107]: Mult-Line Subject is completely broken

**Comment by Flying-Toast** [comment]
Also, headers aren't something that most users are likely to need. I think we should consider adding a mechanism to configure whether or not we want to fetch headers so we avoid downloading excess information from the server. I'm thinking we could somehow extend `Filter`s to do this. Do you have any ideas?

**Comment by ponychicken** [comment]
> I'm not sure I understand the issue. `assert_comms()` replaces all "\n" with "\r\n" then splits on the regex `~r/\r\n[CS]:/`. You should be able to put literal line breaks in the string, as is done [here](https://github.com/Flying-Toast/yugo/blob/3fd2b84e8dbaac739cddb851ed5f6be5541f3c80/test/yugo/client_test.exs#L69). Unless I am missing something - did you try that and it didn't work?

Yes i tried this and it doesn't work. The problem is that both the syntax of the Client and Server response depends on linebreaks for parsing, but also inside the server response there will be linebreaks if you want to add headers, so when assert_comms does the splitting the headers end up being split.

Regarding the other things am not sure i can do this, neither do I feel so good about it. 
I wanted to share my work so that others can profit from it. But I personally don't need the headers parsed for example, just need to archive them. But it seems like the scope of this is growing and growing without clear indication where it goes.

**Comment by ponychicken** [comment]
Ok, thanks. I will have a look at it again next week

**Comment by Flying-Toast** [comment]
Ok. I'll close the PR for now.

**Pull request closed without merging**